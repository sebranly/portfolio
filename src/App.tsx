import React from 'react';
import { Footer } from './components/Footer';
import { projects } from './data/projects';
import { Project } from './components/Project';
import { Language, Project as ProjectType } from './types';
import { useTranslation } from 'react-i18next';
import { languages } from './languages';
import { HeaderText } from './components/HeaderText';
import { Tag } from './types';
import classnames from 'classnames';
import { filterProjectsByTag, generatePages, getProjectsForPage } from './utils';
import { TAB_COUNT } from './constants';

function App() {
  const onClickTab = (tag: Tag) => {
    const newProjects = filterProjectsByTag(projects, tag);
    setApplicationType(tag);
    setSelectedProjects(newProjects);
    setPageNumber(1);
  };

  const onClickPage = (page: number) => {
    setPageNumber(page);
    ref?.current?.scrollIntoView({ behavior: 'smooth' });
  };

  const ref = React.useRef<null | HTMLDivElement>(null);
  const { i18n, t } = useTranslation();
  const [applicationType, setApplicationType] = React.useState(Tag.All);
  const [selectedProjects, setSelectedProjects] = React.useState(projects);
  const [pageNumber, setPageNumber] = React.useState(1);

  const projectsForPage = getProjectsForPage(selectedProjects, pageNumber);
  const allPages = generatePages(selectedProjects.length);

  return (
    <>
      <header className="mx-4 mt-8 mb-4 text-center">
        <div className="mb-4">
          {languages.map((language: Language) => {
            const { shortName, name } = language;
            const isSelected = i18n.resolvedLanguage === shortName;

            return (
              <button
                className="rounded mr-2"
                key={shortName}
                style={{
                  fontWeight: isSelected ? 'bold' : 'normal',
                  textDecorationLine: isSelected ? 'underline' : 'none'
                }}
                type="submit"
                onClick={() => i18n.changeLanguage(shortName)}
              >
                {name}
              </button>
            );
          })}
        </div>
        <HeaderText className="mb-8" />
      </header>
      <div ref={ref} className="mt-4 flex w-full">
        {[Tag.All, Tag.CLI, Tag.GUI, Tag.Website].map((tag: Tag, index: number) => {
          const text = t(`selector.${tag}`);
          const isSelected = applicationType === tag;
          const classnamesTab = classnames(
            {
              'bg-black': isSelected,
              'text-white': isSelected,
              'rounded-tl-lg': isSelected && index !== 0,
              'rounded-tr-lg': isSelected && index !== TAB_COUNT - 1
            },
            `p-2 w-1/${TAB_COUNT}`
          );

          return (
            <button
              className={classnamesTab}
              key={tag}
              style={{ fontWeight: isSelected ? 'bold' : 'normal' }}
              type="submit"
              onClick={() => onClickTab(tag)}
            >
              {text}
            </button>
          );
        })}
      </div>
      <div className="py-4 border border-black flex flex-wrap justify-center bg-black">
        {projectsForPage.map((project: ProjectType) => {
          const { title } = project;
          return <Project key={title} project={project} />;
        })}
      </div>
      <div className="flex w-full">
        {allPages.map((page: number, index: number) => {
          const isSelected = pageNumber === page;
          const classnamesTab = classnames(
            {
              'bg-black': isSelected,
              'text-white': isSelected,
              'rounded-bl-lg': isSelected && index !== 0,
              'rounded-br-lg': isSelected && index !== allPages.length - 1
            },
            `p-2 flex-1`
          );

          return (
            <button
              className={classnamesTab}
              key={page}
              style={{ fontWeight: isSelected ? 'bold' : 'normal' }}
              type="submit"
              onClick={() => onClickPage(page)}
            >
              {page}
            </button>
          );
        })}
      </div>
      <Footer />
    </>
  );
}

export default App;
