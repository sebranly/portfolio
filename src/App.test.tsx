import React from 'react';
import { render } from '@testing-library/react';
import App from './App';

jest.mock("./components/Swiper", () => jest.fn());

test('renders text', () => {
  const { container } = render(<App />);
  expect(container.childNodes).toMatchSnapshot();
});
