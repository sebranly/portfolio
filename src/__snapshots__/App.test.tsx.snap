// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders text 1`] = `
NodeList [
  <header
    class="mx-4 mt-8 mb-4 text-center"
  >
    <div
      class="mb-4"
    >
      <button
        class="rounded mr-2"
        style="font-weight: normal; text-decoration-line: none;"
        type="submit"
      >
        English
      </button>
      <button
        class="rounded mr-2"
        style="font-weight: normal; text-decoration-line: none;"
        type="submit"
      >
        Français
      </button>
    </div>
    <div
      class="mb-8"
    >
      <h1
        class="font-bold md:text-6xl text-2xl"
      >
        header.welcome
      </h1>
      <h2
        class="md:text-4xl text-2xl"
      >
        header.author.prefix
        <div
          class="italic inline"
        >
          sebranly
        </div>
        header.author.suffix
      </h2>
      <div
        class="md:text-2xl text-xl"
      >
        header.introduction
      </div>
    </div>
  </header>,
  <div
    class="mt-4 flex w-full"
  >
    <button
      class="bg-black text-white rounded-tr-lg p-2 w-1/4"
      style="font-weight: bold;"
      type="submit"
    >
      selector.All
    </button>
    <button
      class="p-2 w-1/4"
      style="font-weight: normal;"
      type="submit"
    >
      selector.CLI
    </button>
    <button
      class="p-2 w-1/4"
      style="font-weight: normal;"
      type="submit"
    >
      selector.GUI
    </button>
    <button
      class="p-2 w-1/4"
      style="font-weight: normal;"
      type="submit"
    >
      selector.website
    </button>
  </div>,
  <div
    class="py-4 border border-black flex flex-wrap justify-center bg-black"
  >
    <div
      class="flex flex-col bg-white px-4 py-4 border-2 mx-4 lg:w-2/5 w-full my-2 rounded-lg border-solid border-gray-50"
    >
      <h2
        class="font-bold text-xl"
      >
        projects.ctr2d.title
      </h2>
      <h3
        class="text-lg"
      >
        projects.ctr2d.subtitle
      </h3>
      <div
        class="border-solid border-black border-2 bg-gray-200 p-2 my-2 text-justify"
      >
        <div
          class="mb-2"
        >
          projects.ctr2d.desc1
        </div>
        <div
          class="mb-2"
        >
          projects.ctr2d.desc2
        </div>
        <div
          class="mb-2"
        >
          projects.ctr2d.desc3
        </div>
      </div>
      <div
        class="grow h-4"
      />
      <div>
        <div
          class="border-2 border-black border-solid cursor-pointer inline-block rounded mr-2 my-1 px-4 py-1 text-white bg-black hover:bg-gray-700"
        >
          2011
        </div>
        <div
          class="border-2 border-black border-solid cursor-pointer inline-block rounded mr-2 my-1 px-4 py-1 text-white bg-gray-600 hover:bg-gray-700"
        >
          projects.general.tags.archived
        </div>
        <div
          class="border-2 border-black border-solid cursor-pointer inline-block rounded mr-2 my-1 px-4 py-1 text-white bg-yellow-600 hover:bg-yellow-700"
        >
          C
        </div>
        <div
          class="border-2 border-black border-solid cursor-pointer inline-block rounded mr-2 my-1 px-4 py-1 text-white bg-green-600 hover:bg-green-700"
        >
          GUI
        </div>
        <div
          class="border-2 border-black border-solid cursor-pointer inline-block rounded mr-2 my-1 px-4 py-1 text-white bg-blue-600 hover:bg-blue-700"
        >
          SDL
        </div>
        <div
          class="border-2 border-black border-solid cursor-pointer inline-block rounded mr-2 my-1 px-4 py-1 text-white bg-amber-600 hover:bg-amber-700"
        >
          projects.general.tags.video game
        </div>
        <div
          class="border-2 border-black border-solid cursor-pointer inline-block rounded mr-2 my-1 px-4 py-1 text-white bg-blue-600 hover:bg-blue-700"
        >
          Windows
        </div>
      </div>
    </div>
    <div
      class="flex flex-col bg-white px-4 py-4 border-2 mx-4 lg:w-2/5 w-full my-2 rounded-lg border-solid border-gray-50"
    >
      <h2
        class="font-bold text-xl"
      >
        projects.tetrisSpace.title
      </h2>
      <h3
        class="text-lg"
      >
        projects.tetrisSpace.subtitle
      </h3>
      <div
        class="border-solid border-black border-2 bg-gray-200 p-2 my-2 text-justify"
      >
        <div
          class="mb-2"
        >
          projects.tetrisSpace.desc1
        </div>
        <div
          class="mb-2"
        >
          projects.tetrisSpace.desc2
        </div>
        <div
          class="mb-2"
        >
          projects.tetrisSpace.desc3
        </div>
      </div>
      <div
        class="inline"
      >
        <div
          class="font-bold inline"
        >
          projects.general.contributor.male
          projects.general.colon
           
        </div>
        <div
          class="inline"
        >
          Stéphane (projects.general.roles.Designer, projects.general.roles.Tester)
        </div>
      </div>
      <div
        class="grow h-4"
      />
      <div>
        <div
          class="border-2 border-black border-solid cursor-pointer inline-block rounded mr-2 my-1 px-4 py-1 text-white bg-black hover:bg-gray-700"
        >
          2011
        </div>
        <div
          class="border-2 border-black border-solid cursor-pointer inline-block rounded mr-2 my-1 px-4 py-1 text-white bg-gray-600 hover:bg-gray-700"
        >
          projects.general.tags.archived
        </div>
        <div
          class="border-2 border-black border-solid cursor-pointer inline-block rounded mr-2 my-1 px-4 py-1 text-white bg-yellow-600 hover:bg-yellow-700"
        >
          C
        </div>
        <div
          class="border-2 border-black border-solid cursor-pointer inline-block rounded mr-2 my-1 px-4 py-1 text-white bg-green-600 hover:bg-green-700"
        >
          GUI
        </div>
        <div
          class="border-2 border-black border-solid cursor-pointer inline-block rounded mr-2 my-1 px-4 py-1 text-white bg-red-600 hover:bg-red-700"
        >
          Nintendo DS
        </div>
        <div
          class="border-2 border-black border-solid cursor-pointer inline-block rounded mr-2 my-1 px-4 py-1 text-white bg-orange-600 hover:bg-orange-700"
        >
          PAlib
        </div>
        <div
          class="border-2 border-black border-solid cursor-pointer inline-block rounded mr-2 my-1 px-4 py-1 text-white bg-amber-600 hover:bg-amber-700"
        >
          projects.general.tags.video game
        </div>
      </div>
    </div>
    <div
      class="flex flex-col bg-white px-4 py-4 border-2 mx-4 lg:w-2/5 w-full my-2 rounded-lg border-solid border-gray-50"
    >
      <h2
        class="font-bold text-xl"
      >
        projects.eightQueensPuzzle.title
      </h2>
      <h3
        class="text-lg"
      >
        projects.eightQueensPuzzle.subtitle
      </h3>
      <code
        class="text-black text-sm"
      >
        <pre>
          +---+---+---+---+
        </pre>
        <pre>
          | . | Q | . | . |
        </pre>
        <pre>
          +---+---+---+---+
        </pre>
        <pre>
          | . | . | . | Q |
        </pre>
        <pre>
          +---+---+---+---+
        </pre>
        <pre>
          | Q | . | . | . |
        </pre>
        <pre>
          +---+---+---+---+
        </pre>
        <pre>
          | . | . | Q | . |
        </pre>
        <pre>
          +---+---+---+---+
        </pre>
      </code>
      <div
        class="border-solid border-black border-2 bg-gray-200 p-2 my-2 text-justify"
      >
        <div
          class="mb-2"
        >
          projects.eightQueensPuzzle.desc1
        </div>
        <div
          class="mb-2"
        >
          projects.eightQueensPuzzle.desc2
        </div>
        <div
          class="mb-2"
        >
          projects.eightQueensPuzzle.desc3
        </div>
      </div>
      <div
        class="inline"
      >
        <div
          class="font-bold inline"
        >
          projects.general.code
          projects.general.colon
           
        </div>
        <a
          class="underline inline"
          href="https://github.com/sebranly/eight-queens-puzzle"
          rel="noopener noreferrer"
          target="_blank"
          title="projects.general.altText.github projects.eightQueensPuzzle.title"
        >
          GitHub
        </a>
      </div>
      <div
        class="grow h-4"
      />
      <div>
        <div
          class="border-2 border-black border-solid cursor-pointer inline-block rounded mr-2 my-1 px-4 py-1 text-white bg-gray-600 hover:bg-gray-700"
        >
          2012
        </div>
        <div
          class="border-2 border-black border-solid cursor-pointer inline-block rounded mr-2 my-1 px-4 py-1 text-white bg-gray-600 hover:bg-gray-700"
        >
          projects.general.tags.archived
        </div>
        <div
          class="border-2 border-black border-solid cursor-pointer inline-block rounded mr-2 my-1 px-4 py-1 text-white bg-yellow-600 hover:bg-yellow-700"
        >
          C
        </div>
        <div
          class="border-2 border-black border-solid cursor-pointer inline-block rounded mr-2 my-1 px-4 py-1 text-white bg-black hover:bg-gray-700"
        >
          CLI
        </div>
        <div
          class="border-2 border-black border-solid cursor-pointer inline-block rounded mr-2 my-1 px-4 py-1 text-black bg-white hover:bg-gray-300"
        >
          GitHub
        </div>
        <div
          class="border-2 border-black border-solid cursor-pointer inline-block rounded mr-2 my-1 px-4 py-1 text-white bg-purple-600 hover:bg-purple-700"
        >
          puzzle
        </div>
        <div
          class="border-2 border-black border-solid cursor-pointer inline-block rounded mr-2 my-1 px-4 py-1 text-white bg-amber-600 hover:bg-amber-700"
        >
          projects.general.tags.school project
        </div>
      </div>
    </div>
    <div
      class="flex flex-col bg-white px-4 py-4 border-2 mx-4 lg:w-2/5 w-full my-2 rounded-lg border-solid border-gray-50"
    >
      <h2
        class="font-bold text-xl"
      >
        projects.myDraw.title
      </h2>
      <h3
        class="text-lg"
      >
        projects.myDraw.subtitle
      </h3>
      <div
        class="border-solid border-black border-2 bg-gray-200 p-2 my-2 text-justify"
      >
        <div
          class="mb-2"
        >
          projects.myDraw.desc1
        </div>
        <div
          class="mb-2"
        >
          projects.myDraw.desc2
        </div>
        <div
          class="mb-2"
        >
          projects.myDraw.desc3
        </div>
      </div>
      <div
        class="grow h-4"
      />
      <div>
        <div
          class="border-2 border-black border-solid cursor-pointer inline-block rounded mr-2 my-1 px-4 py-1 text-white bg-gray-600 hover:bg-gray-700"
        >
          2012
        </div>
        <div
          class="border-2 border-black border-solid cursor-pointer inline-block rounded mr-2 my-1 px-4 py-1 text-white bg-gray-600 hover:bg-gray-700"
        >
          projects.general.tags.archived
        </div>
        <div
          class="border-2 border-black border-solid cursor-pointer inline-block rounded mr-2 my-1 px-4 py-1 text-white bg-yellow-600 hover:bg-yellow-700"
        >
          C
        </div>
        <div
          class="border-2 border-black border-solid cursor-pointer inline-block rounded mr-2 my-1 px-4 py-1 text-white bg-green-600 hover:bg-green-700"
        >
          GUI
        </div>
        <div
          class="border-2 border-black border-solid cursor-pointer inline-block rounded mr-2 my-1 px-4 py-1 text-white bg-green-600 hover:bg-green-700"
        >
          myProjects
        </div>
        <div
          class="border-2 border-black border-solid cursor-pointer inline-block rounded mr-2 my-1 px-4 py-1 text-white bg-blue-600 hover:bg-blue-700"
        >
          SDL
        </div>
        <div
          class="border-2 border-black border-solid cursor-pointer inline-block rounded mr-2 my-1 px-4 py-1 text-white bg-blue-600 hover:bg-blue-700"
        >
          Windows
        </div>
      </div>
    </div>
  </div>,
  <div
    class="flex w-full"
  >
    <button
      class="bg-black text-white rounded-br-lg p-2 flex-1"
      style="font-weight: bold;"
      type="submit"
    >
      1
    </button>
    <button
      class="p-2 flex-1"
      style="font-weight: normal;"
      type="submit"
    >
      2
    </button>
    <button
      class="p-2 flex-1"
      style="font-weight: normal;"
      type="submit"
    >
      3
    </button>
    <button
      class="p-2 flex-1"
      style="font-weight: normal;"
      type="submit"
    >
      4
    </button>
    <button
      class="p-2 flex-1"
      style="font-weight: normal;"
      type="submit"
    >
      5
    </button>
    <button
      class="p-2 flex-1"
      style="font-weight: normal;"
      type="submit"
    >
      6
    </button>
    <button
      class="p-2 flex-1"
      style="font-weight: normal;"
      type="submit"
    >
      7
    </button>
  </div>,
  <div
    class="my-4 text-lg text-center"
  >
    <div>
      footer.author.prefix
      <a
        class="underline"
        href="https://github.com/sebranly/portfolio"
        rel="noopener noreferrer"
        target="_blank"
        title="GitHub page for repository"
      >
        sebranly
      </a>
      footer.author.suffix
    </div>
    <div>
      footer.website
      <a
        class="underline"
        href="https://github.com/sebranly/portfolio/releases"
        rel="noopener noreferrer"
        target="_blank"
        title="Website changelog"
      >
        0.0.1
      </a>
    </div>
  </div>,
]
`;
